File: ports/java/src/gui/widgets/BcsWidget.java
package gui.widgets;

import core.kernel.BcsObject;
import java.util.concurrent.locks.ReentrantLock;

public class BcsWidget extends BcsObject {
    private boolean visible = false;
    private boolean enabled = true;
    private int width = 0;
    private int height = 0;
    private int x = 0;
    private int y = 0;
    private final ReentrantLock lock = new ReentrantLock();

    public BcsWidget() {
        this(null);
    }

    public BcsWidget(BcsWidget parent) {
        super(parent);
    }

    public void show() {
        lock.lock();
        try {
            visible = true;
        } finally {
            lock.unlock();
        }
    }

    public void hide() {
        lock.lock();
        try {
            visible = false;
        } finally {
            lock.unlock();
        }
    }

    public boolean isVisible() {
        lock.lock();
        try {
            return visible;
        } finally {
            lock.unlock();
        }
    }

    public void resize(int width, int height) {
        lock.lock();
        try {
            this.width = width;
            this.height = height;
        } finally {
            lock.unlock();
        }
    }

    public void move(int x, int y) {
        lock.lock();
        try {
            this.x = x;
            this.y = y;
        } finally {
            lock.unlock();
        }
    }

    public boolean isEnabled() {
        lock.lock();
        try {
            return enabled;
        } finally {
            lock.unlock();
        }
    }

    public void setEnabled(boolean enabled) {
        lock.lock();
        try {
            this.enabled = enabled;
        } finally {
            lock.unlock();
        }
    }
}

File: ports/go/src/gui/kernel/bcswidget.go
package kernel

import (
	"sync"
	"github.com/robertpelloni/btk-go/ports/go/src/core/kernel"
)

// BcsWidget extends the core BcsWidget mapped in core to include GUI event handlers
type BcsWidget struct {
	*kernel.BcsObject
	visible   bool
	enabled   bool
	width     int
	height    int
	x         int
	y         int
	focus     bool
	mu        sync.RWMutex
}

func NewBcsWidget(parent *kernel.BcsObject) *BcsWidget {
	w := &BcsWidget{
		BcsObject: kernel.NewBcsObject(parent),
		visible:   false,
		enabled:   true,
		width:     0,
		height:    0,
		x:         0,
		y:         0,
		focus:     false,
	}
	return w
}

func (w *BcsWidget) Show() {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.visible = true
}

func (w *BcsWidget) Hide() {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.visible = false
}

func (w *BcsWidget) IsVisible() bool {
	w.mu.RLock()
	defer w.mu.RUnlock()
	return w.visible
}

func (w *BcsWidget) Resize(width, height int) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.width = width
	w.height = height
}

func (w *BcsWidget) Move(x, y int) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.x = x
	w.y = y
}

func (w *BcsWidget) SetEnabled(enabled bool) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.enabled = enabled
}

func (w *BcsWidget) IsEnabled() bool {
	w.mu.RLock()
	defer w.mu.RUnlock()
	return w.enabled
}

func (w *BcsWidget) HasFocus() bool {
	w.mu.RLock()
	defer w.mu.RUnlock()
	return w.focus
}

func (w *BcsWidget) SetFocus(f bool) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.focus = f
}

// Event overrides the base object event router to handle GUI events
func (w *BcsWidget) Event(e kernel.BcsEvent) bool {
	switch e.Type() {
	case kernel.MouseButtonPress:
		w.MousePressEvent(e)
		return true
	case kernel.MouseButtonRelease:
		w.MouseReleaseEvent(e)
		return true
	case kernel.KeyPress:
		w.KeyPressEvent(e)
		return true
	case kernel.KeyRelease:
		w.KeyReleaseEvent(e)
		return true
	case kernel.Paint:
		w.PaintEvent(e)
		return true
	}
	return w.BcsObject.Event(e)
}

func (w *BcsWidget) MousePressEvent(e kernel.BcsEvent) {}
func (w *BcsWidget) MouseReleaseEvent(e kernel.BcsEvent) {}
func (w *BcsWidget) KeyPressEvent(e kernel.BcsEvent) {}
func (w *BcsWidget) KeyReleaseEvent(e kernel.BcsEvent) {}
func (w *BcsWidget) PaintEvent(e kernel.BcsEvent) {}
